proposals/nnnn-isolated-conformances.md

```diff
@@ -0,0 +1,610 @@
+@MainActor func isolatedAnyBad(s: S) {
+  let a: any Sendable & P = s   // error: the (hidden) type parameter for the 'any' is Sendable
+}
+
+@MainActor func returnIsolatedSomeGood(s: S) -> some P {
+  return s   // okay: the 'any P' cannot leave the isolation domain
+}
+
+@MainActor func returnIsolatedSomeBad(s: S) -> some Sendable & P {
+  return s   // error: the (hidden) type parameter for the 'any' is Sendable
+}
+```
+
+As noted in the earlier discussion of data-race safety issues, protecting against values being sent to another isolation domain is insufficient, because passing the *metatype* of the type can also carry the isolated conformances with it. Recall the example from earlier:
+
+```swift
+protocol Q {
+  static func g() { }
+}
+
+nonisolated func callQGElsewhere<T: Q>(_: T.Type) {
+  Task.detached {
+    T.g()
+  }
+}
+```
+
+If the conformance provided for `T: Q` is isolated to a global actor, the call `T.g()` from another isolation domain will break data race safety. This could be diagnosed in the implementation of `callQGElsewhere` if `T.Type` were not guaranteed to be `Sendable`. This sets up a contradiction with SE-0302, which introduced the notion of `Sendable` and specifies that [all metatypes are `Sendable`](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0302-concurrent-value-and-concurrent-closures.md#metatype-conformance-to-sendable). 
+
+We can borrow from the approach taken by [non-copyable](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0427-noncopyable-generics.md) and [non-escapable](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0446-non-escapable.md) types, where we introduce the notion of a suppressed constraint to allow generics to operate with a set of capabilities that is more limited than "normal" types. For example, we could introduce a constraint of the form `T: ~SendableMetatype` to make the metatype of `T` not `Sendable` within a given generic function. If we applied this constraint to `callQGElsewhere`, we would get a compile-time error:
+
+```swift
+nonisolated func callQGElsewhere_restricted<T: Q>(_: T.Type) where T: ~SendableMetatype {
+  Task.detached {
+    T.g() // error: capture of non-Sendable metatype T.Type in concurrently-executed closure
+  }
+}
+```
+
+With this notion, we could amend rule (2) to prohibit using an isolated protocol conformance for a requirement `T: Q` when the generic signature contains either the requirement `T: Sendable` or `T: SendableMetatype`. This closes the data race safety hole with isolated conformances being carried through metatypes.
+
+Unfortunately, this means that isolated conformances won't work with any existing generic code, because all generic code in existence today assumes that all metatypes are `Sendable`. Most of that code could be updated with requirements of the form `T.Type: ~Sendable` and without other changes, but it would require an ecosystem-wide change in support of a somewhat niche feature.
```

Comment:
```suggestion
Unfortunately, this means that isolated conformances won't work with any existing generic code, because all generic code in existence today assumes that all metatypes are `Sendable`. Most of that code could be updated with requirements of the form `T: ~SendableMetatype` and without other changes, but it would require an ecosystem-wide change in support of a somewhat niche feature.
```